Sources/WebSocket/WebSocket.swift

```diff
@@ -0,0 +1,160 @@
+        self.connection = Connection(client: client)
+        var previousType: Frame.OpCode?
+        
+        self.textStream.frameStream = self.connection
+        self.binaryStream.frameStream = self.connection
+        
+        self.connection.drain { frame in
+            func processString() {
+                // invalid string
+                guard let string = frame.data.string() else {
+                    self.connection.client.close()
+                    return
+                }
+                
+                self.textStream.outputStream?(string)
+            }
+            
+            func processBinary() {
+                let readBuffer = ByteBuffer(start: frame.data.baseAddress, count: frame.data.count)
+                
+                self.binaryStream.outputStream?(readBuffer)
+            }
+            
+            switch frame.opCode {
+            case .text:
+                if !frame.final {
+                    previousType = .text
+                }
+                
+                processString()
+            case .binary:
+                if !frame.final {
+                    previousType = .binary
+                }
+                
+                processBinary()
+            case .ping:
+                guard let pointer = frame.data.baseAddress else {
+                    self.connection.client.close()
+                    return
+                }
+                
+                do {
+                    // reply the input
+                    try self.connection.sendFrame(opcode: .pong, pointer: pointer, length: frame.data.count)
+                } catch {
+                    self.connection.errorStream?(error)
+                }
+            case .continuation:
+                defer {
+                    if frame.final {
+                        previousType = nil
+                    }
+                }
+                
+                // TODO: ignore typeless continuations?
+                guard let type = previousType else {
+                    self.connection.client.close()
+                    return
+                }
+                
+                if type == .text {
+                    processString()
+                } else if type == .binary {
+                    processBinary()
+                } else {
+                    // invalid, close or ignore?
+                    self.connection.client.close()
+                    return
+                }
+            case .close:
+                self.connection.client.close()
+            case .pong:
+                return
+            }
+        }
+    }
+}
+
+public class WebSocketMiddleware : RequestMiddleware {
+    public typealias Input = Request
+    public typealias Output = Request
+    
+    public var outputStream: ((Request) -> ())?
+    public var errorStream: BaseStream.ErrorHandler?
+    var websocketStream: ((WebSocket) -> ())?
+    
+    public func onConnect(_ handler: @escaping ((WebSocket) -> ())) {
```

Comment:
This should be its own class like `WebSocketStream` that conforms to `Core.Stream`.